README.md
# Welcome to @biothings-explorer/call-apis 👋

![Test Coveralls](https://github.com/biothings/call-apis.js/workflows/Test%20Coveralls/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/biothings/call-apis.js/badge.svg)](https://coveralls.io/github/biothings/call-apis.js)
<a href="https://github.com/biothings/call-apis.js#readme" target="_blank">
    <img alt="Documentation" src="https://img.shields.io/badge/documentation-yes-brightgreen.svg" />
  </a>
<a href="https://www.npmjs.com/package/@biothings-explorer/call-apis" target="_blank">
    <img alt="Version" src="https://img.shields.io/npm/v/@biothings-explorer/call-apis.svg">
  </a>

A nodejs module to make API calls in batch mode for BioThings Explorer

### 🏠 [Homepage](https://github.com/biothings/call-apis.js)

## Install

```sh
npm i @biothings-explorer/call-apis
```

## Usage

- Import and Initialize

    ```javascript
    const call_apiis = require("@biothings-explorer/call-apis")
    ```

## Run tests

```sh
npm run test
```

## Author

👤 **BioThings team**

* Website: https://biothings.io
* Github: [@biothings](https://github.com/biothings)

## 🤝 Contributing

Contributions, issues and feature requests are welcome!<br />Feel free to check [issues page](https://github.com/biothings/call-apis.js/issues).

## Show your support

Give a ⭐️ if this project helped you!

## 📝 License

Copyright © 2021 [BioThings Team](https://github.com/biothings).<br />
This project is [ISC](https://github.com/biothings/call-apis.js/blob/main/LICENSE) licensed.

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.24.0](https://github.com/biothings/call-apis.js/compare/v1.23.3...v1.24.0) (2021-08-19)

### [1.23.3](https://github.com/biothings/call-apis.js/compare/v1.23.2...v1.23.3) (2021-06-24)


### Bug Fixes

* :bug: ignore undefined items in array ([6b64d09](https://github.com/biothings/call-apis.js/commit/6b64d09a52ab8f4ad469844f977a3d1507d12ccd))

### [1.23.2](https://github.com/biothings/call-apis.js/compare/v1.23.1...v1.23.2) (2021-06-17)


### Bug Fixes

* check  field exists before accessing it ([9b8252f](https://github.com/biothings/call-apis.js/commit/9b8252f6296c4167c83abbbba4a6196cee6e31ec))

### [1.23.1](https://github.com/kevinxin90/call-apis.js/compare/v1.23.0...v1.23.1) (2021-06-04)


### Bug Fixes

* update query message to be compatible with trapi 1.1 ([20b7803](https://github.com/kevinxin90/call-apis.js/commit/20b7803052ae1e964800c9eedfd143113cc7e346))

## [1.23.0](https://github.com/kevinxin90/call-apis.js/compare/v1.22.0...v1.23.0) (2021-05-12)

## [1.22.0](https://github.com/kevinxin90/call-apis.js/compare/v1.21.0...v1.22.0) (2021-05-12)


### Features

* trapi 1.1.0 msg update ([627d54d](https://github.com/kevinxin90/call-apis.js/commit/627d54d469ff8c353a2749a71c6fd9dc2dade33f))

## [1.21.0](https://github.com/kevinxin90/call-apis.js/compare/v1.20.0...v1.21.0) (2021-05-04)

## [1.20.0](https://github.com/kevinxin90/call-apis.js/compare/v1.19.0...v1.20.0) (2021-04-27)


### Bug Fixes

* :bug: change NCBIGENE to NCBIGene, add RHEA ([fa212ff](https://github.com/kevinxin90/call-apis.js/commit/fa212ff5d8bebc1cc1a338f02c83fa98e1afa461))

## [1.19.0](https://github.com/kevinxin90/call-apis.js/compare/v1.18.0...v1.19.0) (2021-04-23)


### Bug Fixes

* :bug: fix issue caused by special character in output ([d690c68](https://github.com/kevinxin90/call-apis.js/commit/d690c68c47e7c64f502765483e59ce9f33ecf1cb))

## [1.18.0](https://github.com/kevinxin90/call-apis.js/compare/v1.17.1...v1.18.0) (2021-04-23)


### Features

* :sparkles: set API timeout limit to be 3s for all TRAPI APIs ([e790fd9](https://github.com/kevinxin90/call-apis.js/commit/e790fd9a256f5a31d3de7276b1675df755479241))
* :sparkles: support biocarta pathway id ([467503c](https://github.com/kevinxin90/call-apis.js/commit/467503cf6d938dd3e9cd3977b6f56efeefe8e345))

### [1.17.1](https://github.com/kevinxin90/call-apis.js/compare/v1.17.0...v1.17.1) (2021-03-30)


### Bug Fixes

* :bug: fix issue when handling reversed trapi response ([4561f58](https://github.com/kevinxin90/call-apis.js/commit/4561f58cf5bc0833c10517a1c6d40ca997bf5ef6))
* :bug: fix issue when looking for bte-trapi tag ([78b23c3](https://github.com/kevinxin90/call-apis.js/commit/78b23c3d24c3dee7161d462a97f682705db0fc2f))

## [1.17.0](https://github.com/kevinxin90/call-apis.js/compare/v1.16.2...v1.17.0) (2021-03-30)


### Features

* :sparkles: supporting calling and retrieving trapi api ([590d329](https://github.com/kevinxin90/call-apis.js/commit/590d3294d0be6eb8698ab0c709ca6f84ce5c7b16))

### [1.16.2](https://github.com/kevinxin90/call-apis.js/compare/v1.16.1...v1.16.2) (2021-03-29)

### [1.16.1](https://github.com/kevinxin90/call-apis.js/compare/v1.16.0...v1.16.1) (2021-03-26)


### Bug Fixes

* :bug: not deepcopy edge_metadata ([59bd6b7](https://github.com/kevinxin90/call-apis.js/commit/59bd6b77b6787df3fcac8f9c70d2c5578e2de51a))

## [1.16.0](https://github.com/kevinxin90/call-apis.js/compare/v1.15.3...v1.16.0) (2021-03-26)

### [1.15.3](https://github.com/kevinxin90/call-apis.js/compare/v1.15.2...v1.15.3) (2021-03-26)

### [1.15.2](https://github.com/kevinxin90/call-apis.js/compare/v1.15.1...v1.15.2) (2021-03-25)

### [1.15.1](https://github.com/kevinxin90/call-apis.js/compare/v1.15.0...v1.15.1) (2021-03-25)

## [1.15.0](https://github.com/kevinxin90/call-apis.js/compare/v1.14.0...v1.15.0) (2021-03-24)

## [1.14.0](https://github.com/kevinxin90/call-apis.js/compare/v1.13.0...v1.14.0) (2021-03-18)


### Features

* :sparkles: add num_of_participants property for pathway ([5800427](https://github.com/kevinxin90/call-apis.js/commit/580042787b6623528784cabe973d7082b2080862))

## [1.13.0](https://github.com/kevinxin90/call-apis.js/compare/v1.12.1...v1.13.0) (2021-03-18)


### Features

* :sparkles: add additional node attributes for resolved ids ([5bb9510](https://github.com/kevinxin90/call-apis.js/commit/5bb9510fa41e62dba0a210ee6a65dd34c8a51a5d))

### [1.12.1](https://github.com/kevinxin90/call-apis.js/compare/v1.12.0...v1.12.1) (2021-03-12)

## [1.12.0](https://github.com/kevinxin90/call-apis.js/compare/v1.11.1...v1.12.0) (2021-03-02)


### Features

* :sparkles: support traversing biolink hierarchy when resolving ids ([204bf3e](https://github.com/kevinxin90/call-apis.js/commit/204bf3e7489282935397e6261b58bacfc39afd70))

### [1.11.1](https://github.com/kevinxin90/call-apis.js/compare/v1.11.0...v1.11.1) (2021-02-22)

## [1.11.0](https://github.com/kevinxin90/call-apis.js/compare/v1.10.5...v1.11.0) (2021-02-18)


### Features

* :sparkles: generate invalidBioEntities if enable ID resolution set to false ([87ce0ab](https://github.com/kevinxin90/call-apis.js/commit/87ce0ab634a5ea75785c5a4b5b038233150fda3a))

### [1.10.5](https://github.com/kevinxin90/call-apis.js/compare/v1.10.4...v1.10.5) (2021-02-18)


### Bug Fixes

* :bug: fix issue that getConfig is called multiple times ([5f0f697](https://github.com/kevinxin90/call-apis.js/commit/5f0f697f17c3b367ef6905ea00eb81ad1c99e340))

### [1.10.4](https://github.com/kevinxin90/call-apis.js/compare/v1.10.3...v1.10.4) (2021-02-17)


### Bug Fixes

* :bug: fix wrong axios config ([5e29c99](https://github.com/kevinxin90/call-apis.js/commit/5e29c99114accb4c399d873d1b52965c6051668f))

### [1.10.3](https://github.com/kevinxin90/call-apis.js/compare/v1.10.2...v1.10.3) (2021-02-14)


### Bug Fixes


kevin.js
const qb = require("./query_builder");
const axios = require("axios");
const qd = require("./index");

const edge = {
    "input": ["C1332824", "C1332823", "123"],
    "query_operation": {
        "params": {
            "fields": "positively_regulates"
        },
        "request_body": {
            "body": {
                "q": "{inputs[0]}",
                "scopes": "umls"
            }
        },
        "path": "/query",
        "path_params": [],
        "method": "post",
        "server": "https://biothings.ncats.io/semmedgene",
        "tags": [
            "disease",
            "annotation",
            "query",
            "translator",
            "biothings",
            "semmed"
        ],
        "supportBatch": true,
        "inputSeparator": ","
    },
    "association": {
        "input_id": "UMLS",
        "input_type": "Gene",
        "output_id": "UMLS",
        "output_type": "Gene",
        "predicate": "positively_regulates",
        "source": "SEMMED",
        "api_name": "SEMMED Gene API",
        "smartapi": {
            "id": "81955d376a10505c1c69cd06dbda3047",
            "meta": {
                "ETag": "f94053bc78b3c2f0b97f7afd52d7de2fe083b655e56a53090ad73e12be83673b",
                "github_username": "kevinxin90",
                "timestamp": "2020-05-27T16:53:40.804575",
                "uptime_status": "good",
                "uptime_ts": "2020-06-12T00:04:31.404599",
                "url": "https://raw.githubusercontent.com/NCATS-Tangerine/translator-api-registry/master/semmed/semmed_gene.yaml"
            }
        }
    },
    "response_mapping": {
        "positively_regulates": {
            "pmid": "positively_regulates.pmid",
            "umls": "positively_regulates.umls"
        }
    },
    "id": "01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b"
};

const edge1 = {
    "input": ["223"],
    "query_operation": {
        "params": {
            "fields": "associated_with"
        },
        "request_body": {
            "body": {
                "q": "{inputs[0]}",
                "scopes": "hgnc"
            },
            "header": "application/x-www-form-urlencoded"
        },
        "path": "/query",
        "path_params": [],
        "method": "post",
        "server": "https://biothings.ncats.io/cord_gene",
        "tags": [
            "gene",
            "annotation",
            "query",
            "translator",
            "biothings"
        ],
        "supportBatch": true,
        "inputSeparator": ","
    },
    "association": {
        "input_id": "HGNC",
        "input_type": "Gene",
        "output_id": "DOID",
        "output_type": "Disease",
        "predicate": "related_to",
        "source": "Translator Text Mining Provider",
        "api_name": "CORD Gene API",
        "smartapi": {
            "id": "6bc54230a6fa7693b2cd113430387ca7",
            "meta": {
                "ETag": "5e7512d15d24b57b52cb15604aaa6c24192f48ef00da9732f23aab3707b2061b",
                "github_username": "kevinxin90",
                "timestamp": "2020-04-29T00:00:40.725359",
                "uptime_status": "good",
                "uptime_ts": "2020-06-12T00:05:25.251375",
                "url": "https://raw.githubusercontent.com/NCATS-Tangerine/translator-api-registry/master/cord/cord_gene.yml"
            }
        }
    },
    "response_mapping": {
        "related_to": {
            "HGNC": "associated_with.hgnc",
            "pmc": "associated_with.pmc"
        }
    },
    "id": "01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b"
};

const edge2 = {
    "input": ["238", "239", "240"],
    "query_operation": {
        "params": {
            "fields": "associated_with"
        },
        "request_body": {
            "body": {
                "q": "{inputs[0]}",
                "scopes": "hgnc"
            },
            "header": "application/x-www-form-urlencoded"
        },
        "path": "/query",
        "path_params": [],
        "method": "post",
        "server": "https://biothings.ncats.io/cord_gene",
        "tags": [
            "gene",
            "annotation",
            "query",
            "translator",
            "biothings"
        ],
        "supportBatch": true,
        "inputSeparator": ","
    },
    "association": {
        "input_id": "HGNC",
        "input_type": "Gene",
        "output_id": "HGNC",
        "output_type": "Gene",
        "predicate": "related_to",
        "source": "Translator Text Mining Provider",
        "api_name": "CORD Gene API",
        "smartapi": {
            "id": "6bc54230a6fa7693b2cd113430387ca7",
            "meta": {
                "ETag": "5e7512d15d24b57b52cb15604aaa6c24192f48ef00da9732f23aab3707b2061b",
                "github_username": "kevinxin90",
                "timestamp": "2020-04-29T00:00:40.725359",
                "uptime_status": "good",
                "uptime_ts": "2020-06-12T00:05:25.251375",
                "url": "https://raw.githubusercontent.com/NCATS-Tangerine/translator-api-registry/master/cord/cord_gene.yml"
            }
        }
    },
    "response_mapping": {
        "related_to": {
            "HGNC": "associated_with.hgnc",
            "pmc": "associated_with.pmc"
        }
    },
    "id": "01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b"
};




const qqq = async () => {
    let qd1 = new qd([edge, edge1, edge2]);
    let res = await qd1.query();
    console.log(qd1.result.slice(-1)[0]);
}

const q = async () => {
    let res = await axios(a.config);
    console.log(res.data);
}

const qq = async () => {
    let res = await axios({
        method: 'post',
        url: 'https://biothings.ncats.io/semmedgene/query',
        data: 'q=C1332823, C1332824, 123&scopes=umls',
        params: {
            fields: 'name,umls,positively_regulates',
            size: '5'
        }
    });
    console.log(res.data);
};

qqq();